Add support for custom compiler passes

ubick opened the pull request:
- [x] Add support for custom compiler passes

```
Dispatch custom "symfony_container_before_container_generator" Magento event
before container compiler passes are registered.

This will allow any Magento observer listening to the above event to inject
custom compiler passes into the configuration object just before the
container is created.
```
- [x]   Delete cache meta file during cache clear

```
The latest versin of this library also creates a "container.cache.php.meta"
file when the container is cached.
Updated the Magento observer to also remove this file if it exists when
the Magento cache clear event is triggered.
```
## How to register a custom compiler pass?
- create a custom compiler pass class in your project (e.g. https://github.com/inviqa/magento-symfony-container/blob/master/app/code/community/Inviqa/SymfonyContainer/Model/InjectableCompilerPass.php)
- create a Magento observer and configure it to listen to the `symfony_container_before_container_generator` Magento event
- update the above observer to inject the custom compiler pass into the generator config, as shown below:

``` php
class Inviqa_DependencyInjection_Model_Observer
{
    /** @var CustomCompilerPass */
    private $customCompilerPass;

    public function __construct()
    {
        $this->customCompilerPass = new CustomCompilerPass();
    }

    public function onBeforeContainerGenerator(Varien_Event_Observer $observer)
    {
        /** @var Configuration $generatorConfig */
        $generatorConfig = $observer->getData('generator_config');

        $generatorConfig->addCompilerPass($this->customCompilerPass);
    }
}
```

Commit:
Add support for custom compiler passes

Dispatch custom "symfony_container_before_container_generator" Magento event
before container compiler passes are registered.

This will allow any Magento observer listening to the above event to inject
custom compiler passes into the configuration object just before the
container is created.

Commit:
Delete cache meta file during cache clear

The latest versin of this library also creates a "container.cache.php.meta"
file when the container is cached.
Updated the Magento observer to also remove this file if it exists when
the Magento cache clear event is triggered.

adragus-inviqa commented on app/code/community/Inviqa/SymfonyContainer/Helper/ContainerProvider.php:
```diff
@@ -22,14 +29,15 @@ class Inviqa_SymfonyContainer_Helper_ContainerProvider
 
     public function __construct(array $services = array())
     {
+        $this->_mageApp = isset($services['app']) ? $services['app'] : Mage::app();
```
`instanceof` check maybe?

ubick replied:
Is the suggestion around replacing `isset(..)` with `instanceof(..)`? What type would `instanceof` check against? The interface `MageApp`? What would be the advantage doing that?

`isset` was used as that looks to be the standard way of checking for injectable dependencies in this project.

adragus-inviqa replied:
But isn't `$_mageApp` assumed to be `Mage_Core_Model_App`?

mlopes commented on app/code/community/Inviqa/SymfonyContainer/Model/Observer.php:
```diff
@@ -12,9 +12,15 @@ class Inviqa_SymfonyContainer_Model_Observer
-            $filePath = Mage::getBaseDir('cache') . '/' . ConfigurationBuilder::CACHED_CONTAINER;
-            if (file_exists($filePath)) {
-                unlink($filePath);
+            $containerFilePath = Mage::getBaseDir('cache') . '/' . ConfigurationBuilder::CACHED_CONTAINER;
```
Should this use `DIRECTORY_SEPARATOR` so that it can be ported to other systems more easily?

adragus-inviqa replied:
Every OS has support for fw-slashes. When _reading_ paths you'd need `DIRECTORY_SEPARATOR`. So this is fine.

jon-acker replied:
@adragus-inviqa I was also thinking we should be using DIRECTORY_SEPARATOR, doesn't file_exists() involve reading?

adragus-inviqa replied:
It does, @jon-acker, but you read _a file_ (or its properties), whereas I meant when you read a _path_ that the system gives _you_.

mlopes commented on app/code/community/Inviqa/SymfonyContainer/Model/Observer.php:
```diff
@@ -12,9 +12,15 @@ class Inviqa_SymfonyContainer_Model_Observer
+            $containerFilePath = Mage::getBaseDir('cache') . '/' . ConfigurationBuilder::CACHED_CONTAINER;
+            $metaFilePath = Mage::getBaseDir('cache') . '/' . ConfigurationBuilder::CACHED_CONTAINER . '.meta';
+
+            if (file_exists($containerFilePath)) {
```
Maybe one more for discussion rather than a change suggestion:

Would probably extract these as there's clear duplication. On the other hand this as an extracted method is really not what I would expect to find in this class, and it is possible that the act of deleting the file is not meaningful enough in the context to deserve its own abstraction.

jon-acker commented on app/code/spec/Inviqa/SymfonyContainer/Helper/ContainerProviderSpec.php:
```diff
@@ -30,18 +37,28 @@ function let(Configuration $generatorConfig, StoreConfigCompilerPass $configComp
     }
 
-    function it_memoizes_container(Configuration $generatorConfig, StoreConfigCompilerPass $configCompilerPass, InjectableCompilerPass $injectableCompilerPass)
+    function it_memorizes_the_container(
```
@ubick it actually is memoize https://en.wikipedia.org/wiki/Memoization 😸

jon-acker commented:
Looks good, thanks.
and 👍 for the caching fix.

ubick commented:
Thanks for the helpful feedback everyone. I'll get back to this PR in the upcoming days.

Commit:
Refactor container observer

Replace forward slash with DIRECTORY_SEPARATOR
Extract the cache file path to private methods

Commit:
Fix memoize spec name
